server/util/fsync/fsync.go

```diff
@@ -0,0 +1,142 @@
+package fsync
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"golang.org/x/sys/unix"
+)
+
+// SyncPath opens a path and syncs it to disk.
+func SyncPath(path string) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return fmt.Errorf("open path for sync: %w", err)
+	}
+	defer f.Close()
+	return f.Sync()
+}
+
+// SyncParentDir syncs the parent directory of a path to disk.
+func SyncParentDir(path string) error {
+	return SyncPath(filepath.Dir(path))
+}
+
+// MkdirAllAndSync creates a directory tree and syncs all created directories and their parents to disk.
+func MkdirAllAndSync(path string, mode os.FileMode) error {
```

Comment:
hm this feels slightly heavy-handed. We're syncing all the dirs up to the root of the filesystem (`/`) instead of just up to the root of the downloaded layer dir. Could maybe take two path parameters, one is a "root" dir at which we should stop recursing upward after syncing it, the other is a path relative to that dir?